[SPARK-32573][SQL] Anti Join Improvement with EmptyHashedRelation and EmptyHashedRelationWithAllNullKeys

leanken-zz opened the pull request:
### What changes were proposed in this pull request?
In [SPARK-32290](https://issues.apache.org/jira/browse/SPARK-32290), we introduced several new types of HashedRelation.

* EmptyHashedRelation
* EmptyHashedRelationWithAllNullKeys

They were all limited to used only in NAAJ scenario. These new HashedRelation could be applied to other scenario for performance improvements.

* EmptyHashedRelation could also be used in Normal AntiJoin for fast stop
* While AQE is on and buildSide is EmptyHashedRelationWithAllNullKeys, can convert NAAJ to a Empty LocalRelation to skip meaningless data iteration since in Single-Key NAAJ, if null key exists in BuildSide, will drop all records in streamedSide.

This Patch including two changes.

* using EmptyHashedRelation to do fast stop for common anti join as well
* In AQE, eliminate BroadcastHashJoin(NAAJ) if buildSide is a EmptyHashedRelationWithAllNullKeys

### Why are the changes needed?
LeftAntiJoin could apply `fast stop` when BuildSide is EmptyHashedRelation, While within AQE with EmptyHashedRelationWithAllNullKeys, we can eliminate the NAAJ. This should be a performance improvement in AntiJoin.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?

* added case in AdaptiveQueryExecSuite.
* added case in HashedRelationSuite.
* Make sure SubquerySuite JoinSuite SQLQueryTestSuite passed.

leanken-zz committed:
[SPARK-32573][SQL] Eliminate Anti Join when BuildSide is Empty

In [SPARK-32290](https://issues.apache.org/jira/browse/SPARK-32290), we introduced several new types of HashedRelation.

* EmptyHashedRelation
* EmptyHashedRelationWithAllNullKeys

They were all limited to used only in NAAJ scenario. But as for a improvement, EmptyHashedRelation could also be used in Normal AntiJoin for fast stop, and as for in AQE, we can even eliminate anti join when we knew that buildSide is empty.

This Patch including two changes.

* In Non-AQE, using EmptyHashedRelation to do fast stop for common anti join as well
* In AQE, eliminate anti join if buildSide is a EmptyHashedRelation of ShuffleWriteRecord is 0

LeftAntiJoin could apply `fast stop` when BuildSide is Empty, While within AQE, we can even eliminate the anti join. This should be a performance improvement in AntiJoin.

No.

* added case in AdaptiveQueryExecSuite.
* added case in HashedRelationSuite.
* Make sure SubquerySuite JoinSuite SQLQueryTestSuite passed.

Change-Id: I718227725e319a54b99c91e6a0d1b9f022343b16

leanken-zz committed:
remove useless import.

Change-Id: I47ebe0cc2e751b018a6ee02fc3018601805c4060

leanken commented:
@cloud-fan could you please have a look at this followup?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

maropu commented:
ok to test

SparkQA commented:
**[Test build #127231 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127231/testReport)** for PR 29389 at commit [`25dd444`](https://github.com/apache/spark/commit/25dd44417e118ea87470c00ed11393055eb2ee52).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31851/
Test PASSed.

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateAntiJoin.scala:
```diff
@@ -0,0 +1,44 @@
+ */
+case class EliminateAntiJoin(conf: SQLConf) extends Rule[LogicalPlan] {
+
+  private def shouldEliminate(plan: LogicalPlan): Boolean = plan match {
```
nit: `should` -> `can`?

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala:
```diff
@@ -82,6 +82,8 @@ case class BroadcastExchangeExec(
     "buildTime" -> SQLMetrics.createTimingMetric(sparkContext, "time to build"),
     "broadcastTime" -> SQLMetrics.createTimingMetric(sparkContext, "time to broadcast"))
 
+  var zeroNumRows: Boolean = false
```
nit: How about rewriting it like this?
```
+  private var knownRowCount: Option[BigInt] = None
+
   override def runtimeStatistics: Statistics = {
     val dataSize = metrics("dataSize").value
-    Statistics(dataSize, rowCount = if (zeroNumRows) Some(0L) else None)
+    Statistics(dataSize, rowCount = knownRowCount)
   }
 
   @transient
@@ -128,6 +128,9 @@ case class BroadcastExchangeExec(
             val relation = mode.transform(input, Some(numRows))
 
             val dataSize = relation match {
+              case EmptyHashedRelation =>
+                knownRowCount = Some(0L)
+                0L
               case map: HashedRelation =>
                 map.estimatedSize
               case arr: Array[InternalRow] =>
@@ -137,7 +140,6 @@ case class BroadcastExchangeExec(
                   s"type: ${relation.getClass.getName}")
             }
 
-            zeroNumRows = relation == EmptyHashedRelation

```

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala:
```diff
@@ -125,7 +125,7 @@ private[execution] object HashedRelation {
  *  [size of key] [size of value] [key bytes] [bytes for value]
  */
-private[joins] class UnsafeHashedRelation(
+class UnsafeHashedRelation(
```
We need this change?

leanken replied:
Ah. this is historical code. A question, should we consider the same when
LongHashedRelation.keys.isEmpty
UnsafeHashedRelation.keys.isEmpty
equalTo EmptyHashedRelation
if it does, we need this private[joins] to be removed. I am not quite sure.

maropu replied:
Ah, I see. it is a trivial change and it looks okay.

leanken-zz committed:
Nit & Code Style.

Change-Id: Icfe57ed02de8dd1b414f4d0c81a51353d49a6ead

leanken commented:
@maropu updated. thanks for doing this in weekend. ^_^

SparkQA commented:
**[Test build #127233 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127233/testReport)** for PR 29389 at commit [`2063288`](https://github.com/apache/spark/commit/2063288e072a266218b57772ada874d35432b344).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31853/
Test PASSed.

maropu commented:
You, too, @leanken ;) thanks for the work, anyway.

SparkQA commented:
**[Test build #127231 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127231/testReport)** for PR 29389 at commit [`25dd444`](https://github.com/apache/spark/commit/25dd44417e118ea87470c00ed11393055eb2ee52).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127231/
Test PASSed.

SparkQA commented:
**[Test build #127233 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127233/testReport)** for PR 29389 at commit [`2063288`](https://github.com/apache/spark/commit/2063288e072a266218b57772ada874d35432b344).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127233/
Test PASSed.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateAntiJoin.scala:
```diff
@@ -0,0 +1,44 @@
+    // If the right side is empty, LeftAntiJoin simply returns the left side.
+    // Eliminate Join with left LogicalPlan instead.
+    case Join(left, right, LeftAnti, _, _)
+        if canEliminate(right) =>
```
nit. Shall we make this one line, `case Join(left, right, LeftAnti, _, _) if canEliminate(right) =>`?

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:
```diff
@@ -223,19 +223,19 @@ case class BroadcastHashJoinExec(
+         |// If the right side is empty, AntiJoin simply returns the left side.
+         |$numOutput.add(1);
+         |${consume(ctx, input)}
+         """.stripMargin
```
indentation?

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala:
```diff
@@ -950,8 +950,18 @@ trait NullAwareHashedRelation extends HashedRelation with Externalizable {
 /**
  * A special HashedRelation indicates it built from a empty input:Iterator[InternalRow].
+ * get & getValue will return null just like
+ * Empty LongHashedRelation or Empty UnsafeHashedRelation does.
```
`Empty` -> `empty`

leanken-zz committed:
nit & indentation.

Change-Id: If5d776916fc12674068c486d605ff9a1fc28526f

leanken commented:
@dongjoon-hyun updated, thanks

SparkQA commented:
**[Test build #127249 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127249/testReport)** for PR 29389 at commit [`257745a`](https://github.com/apache/spark/commit/257745ae314de2ea7a186916aa7b47ce66a60ca9).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31869/
Test PASSed.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala:
```diff
@@ -82,6 +82,8 @@ case class BroadcastExchangeExec(
     "buildTime" -> SQLMetrics.createTimingMetric(sparkContext, "time to build"),
     "broadcastTime" -> SQLMetrics.createTimingMetric(sparkContext, "time to broadcast"))
 
+  private var knownRowCount: Option[BigInt] = None
```
can we be consistent and use `metrics` to hold rowCount? You can follow how we track `dataSize`.

leanken replied:
Sure will do the update.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala:
```diff
@@ -126,6 +128,9 @@ case class BroadcastExchangeExec(
 
             val dataSize = relation match {
+              case EmptyHashedRelation =>
+                knownRowCount = Some(0L)
```
See L118. We already know the row count and we even use it to check `numRows >= MAX_BROADCAST_TABLE_ROWS`

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala:
```diff
@@ -950,8 +950,18 @@ trait NullAwareHashedRelation extends HashedRelation with Externalizable {
+ * empty LongHashedRelation or empty UnsafeHashedRelation does.
  */
 object EmptyHashedRelation extends NullAwareHashedRelation {
+  override def get(key: Long): Iterator[InternalRow] = null
```
are you sure `UnsafeHashedRelation` returns null not `Iterator.empty`?

leanken replied:
```
override def get(key: InternalRow): Iterator[InternalRow] = {
    val unsafeKey = key.asInstanceOf[UnsafeRow]
    val map = binaryMap  // avoid the compiler error
    val loc = new map.Location  // this could be allocated in stack
    binaryMap.safeLookup(unsafeKey.getBaseObject, unsafeKey.getBaseOffset,
      unsafeKey.getSizeInBytes, loc, unsafeKey.hashCode())
    if (loc.isDefined) {
      new Iterator[UnsafeRow] {
        private var _hasNext = true
        override def hasNext: Boolean = _hasNext
        override def next(): UnsafeRow = {
          resultRow.pointTo(loc.getValueBase, loc.getValueOffset, loc.getValueLength)
          _hasNext = loc.nextValue()
          resultRow
        }
      }
    } else {
      null // return null if not match
    }
  }

  def getValue(key: InternalRow): InternalRow = {
    val unsafeKey = key.asInstanceOf[UnsafeRow]
    val map = binaryMap  // avoid the compiler error
    val loc = new map.Location  // this could be allocated in stack
    binaryMap.safeLookup(unsafeKey.getBaseObject, unsafeKey.getBaseOffset,
      unsafeKey.getSizeInBytes, loc, unsafeKey.hashCode())
    if (loc.isDefined) {
      resultRow.pointTo(loc.getValueBase, loc.getValueOffset, loc.getValueLength)
      resultRow
    } else {
      null // return null if not match
    }
  }
```

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala:
```diff
@@ -950,8 +950,18 @@ trait NullAwareHashedRelation extends HashedRelation with Externalizable {
 
 /**
  * A special HashedRelation indicates it built from a empty input:Iterator[InternalRow].
+ * get & getValue will return null just like
```
BTW where do we call these methods for `EmptyHashedRelation`?

leanken replied:
yes. It will be called. EmptyHashedRelation now is also applied at all joinType. I confirmed it while running UT.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:
```diff
@@ -223,19 +223,19 @@ case class BroadcastHashJoinExec(
-           |${consume(ctx, input)}
-         """.stripMargin
-      } else if (broadcastRelation.value == EmptyHashedRelationWithAllNullKeys) {
+      if (broadcastRelation.value == EmptyHashedRelationWithAllNullKeys) {
```
shall we handle this case in AQE as well?

leanken replied:
I think EmptyHashedRelationWithAllNullKeys should only be handled in NAAJ scenario. It might be needed to consider in my next follow up => support NAAJ in ShuffleHashJoinExec.

leanken replied:
do you mean to eliminate NAAJ BroadcastJoin if it is a EmptyHashedRelationWithAllNullKeys?

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateAntiJoin.scala:
```diff
@@ -0,0 +1,43 @@
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformDown {
+    // If the right side is empty, LeftAntiJoin simply returns the left side.
+    // Eliminate Join with left LogicalPlan instead.
+    case Join(left, right, LeftAnti, _, _) if canEliminate(right) =>
```
This is not very useful, as the anti join is already very fast with the empty right side.

https://github.com/apache/spark/pull/29389/files#r467683621 is more useful, as for now we still need to consume all the left side data and return nothing. It will be a big improvement if we can directly convert the join node to an empty relation with AQE.

leanken replied:
> This is not very useful, as the anti join is already very fast with the empty right side.
> 
> https://github.com/apache/spark/pull/29389/files#r467683621 is more useful, as for now we still need to consume all the left side data and return nothing. It will be a big improvement if we can directly convert the join node to an empty relation with AQE.

In fact, we need to consume all the left side data and return all the rows when it's an EmptyHashedRelation.

cloud-fan replied:
> In fact, we need to consume all the left side data and return all the rows when it's an EmptyHashedRelation.

Yes, but this rule doesn't change it. By returning the left side plan, we still need to consume all the data of the left side.

leanken-zz committed:
eliminate normal anti join when EmptyHashedRelation is not neccesary.
eliminate naaj which buildSide is EmptyHashedRelationWithAllNullKeys while AQE is on, should gain more Performance improvement.

Change-Id: I13c667628220da923fc59718281642460c80f71d

leanken-zz committed:
file renamed to EliminateNullAwareAntiJoin since it's for NAAJ.

Change-Id: I0d0a96da107145b9fc681e15f2533abd165f344a

leanken-zz committed:
remove SQLConf in EliminateNullAwareAntiJoin.

Change-Id: I3039efd9e472431131b52bec0d0619b84e2c6ee5

SparkQA commented:
**[Test build #127249 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127249/testReport)** for PR 29389 at commit [`257745a`](https://github.com/apache/spark/commit/257745ae314de2ea7a186916aa7b47ce66a60ca9).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

leanken commented:
@cloud-fan updated.

leanken commented:
1. keep EmptyHashedRelation for common anti join fast stop both for NON-AQE and AQE
2. rewrite NAAJ to Empty LocalRelation when BuildSide is EmptyHashedRelationWithAllNullKeys with AQE on.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127249/
Test FAILed.

SparkQA commented:
**[Test build #127270 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127270/testReport)** for PR 29389 at commit [`381db11`](https://github.com/apache/spark/commit/381db1115c1543513c77369306ca11a3ce00ad47).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31890/
Test PASSed.

cloud-fan commented:
@leanken please also update the PR description.

leanken commented:
Update PR title => Anti Join Improvement with EmptyHashedRelation and EmptyHashedRelationWithAllNullKeys

leanken commented:
Update RP desc and Jira desc accordingly, @cloud-fan

SparkQA commented:
**[Test build #127270 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127270/testReport)** for PR 29389 at commit [`381db11`](https://github.com/apache/spark/commit/381db1115c1543513c77369306ca11a3ce00ad47).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127270/
Test PASSed.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala:
```diff
@@ -223,19 +223,19 @@ case class BroadcastHashJoinExec(
   protected override def codegenAnti(ctx: CodegenContext, input: Seq[ExprCode]): String = {
-    if (isNullAwareAntiJoin) {
-      val (broadcastRelation, relationTerm) = prepareBroadcast(ctx)
+    val (broadcastRelation, relationTerm) = prepareBroadcast(ctx)
```
After a second look, I think this is incorrect. We may go to the `else` branch, which calls `super.codegenAnti` and call `prepareBroadcast` again.

I think we should keep this method untouched, and update `HashJoin.codegenAnti` to fast stop with `EmptyHashedRelation`.

leanken replied:
I thought about this problem too.
HashJoin.scala does not have the API prepareBroadcast, only the following prepareRelation available, which will not return HashedRelation.

```
protected def prepareRelation(ctx: CodegenContext): (String, Boolean)
```

I was thinking that call prepareBroadcast twice is OK since broadcast.Broadcast[HashedRelation] is a future call, and the collect and build action will only happen once, but the relationTerm will be generated twice

is it ok to just change prepareRelation to return a (String, Boolean, HashedRelation) ??
tell me when you decided which way to go. @cloud-fan

leanken replied:
I was also consider doing it inside BHJ because in ShuffledHashJoinExec
the hashedRelation is built among generated-code, it might be hard to just return hashedRelation in prepareRelation call.

```
protected override def prepareRelation(ctx: CodegenContext): (String, Boolean) = {
    val thisPlan = ctx.addReferenceObj("plan", this)
    val clsName = classOf[HashedRelation].getName

    // Inline mutable state since not many join operations in a task
    val relationTerm = ctx.addMutableState(clsName, "relation",
      v => s"$v = $thisPlan.buildHashedRelation(inputs[1]);", forceInline = true)
    (relationTerm, false)
  }
```

leanken-zz committed:
refine prepareRelation to avoid call prepareBroadcast twice.

Change-Id: Ieb631235d4bc8f1446387eb22f0c1a1cc1d3ed27

SparkQA commented:
**[Test build #127286 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127286/testReport)** for PR 29389 at commit [`c8eae30`](https://github.com/apache/spark/commit/c8eae30493ba7df0c5918a00d0b1c58c681b27ca).

leanken commented:
@cloud-fan updated. refine prepareRelation to return

```
private[joins] case class HashedRelationInfo(
    relationTerm: String,
    keyIsUnique: Boolean,
    isEmptyHashedRelation: Boolean)
```

Code change int BHJ reverted, and do fast stop in HashJoin.scala codegenAnti using isEmptyHashedRelation

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31906/
Test PASSed.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala:
```diff
@@ -29,6 +29,16 @@ import org.apache.spark.sql.execution.{CodegenSupport, ExplainUtils, RowIterator
+private[joins] case class HashedRelationInfo(
+    relationTerm: String,
+    keyIsUnique: Boolean,
+    isEmptyHashedRelation: Boolean)
```
we can probably make the name shorter, just use `isEmpty`.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala:
```diff
@@ -81,13 +81,13 @@ case class ShuffledHashJoinExec(
     val relationTerm = ctx.addMutableState(clsName, "relation",
       v => s"$v = $thisPlan.buildHashedRelation(inputs[1]);", forceInline = true)
-    (relationTerm, false)
+    HashedRelationInfo(relationTerm, false, false)
```
nit: `...(..., isKeyUnique = false, isEmpty = false)`.

cloud-fan reviewed (approved):
LGTM except a few nits.

SparkQA commented:
**[Test build #127286 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127286/testReport)** for PR 29389 at commit [`c8eae30`](https://github.com/apache/spark/commit/c8eae30493ba7df0c5918a00d0b1c58c681b27ca).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127286/
Test PASSed.

leanken-zz committed:
nit.

Change-Id: I726afddeea0c49d5299b7dc9e8e27f0021411418

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/31915/
Test PASSed.

SparkQA commented:
**[Test build #127296 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127296/testReport)** for PR 29389 at commit [`5309b66`](https://github.com/apache/spark/commit/5309b66b23af2ffe7999edbd8f0c11f26628805d).

SparkQA commented:
**[Test build #127296 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/127296/testReport)** for PR 29389 at commit [`5309b66`](https://github.com/apache/spark/commit/5309b66b23af2ffe7999edbd8f0c11f26628805d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

leanken commented:
@cloud-fan Test passed, ready to merge.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/127296/
Test PASSed.

cloud-fan commented:
thanks, merging to master!
